consensus/src/block_storage/sync_manager.rs

```diff
@@ -259,54 +287,95 @@ impl BlockRetriever {
     /// error is returned
-    async fn retrieve_block_for_qc<'a>(
-        &'a mut self,
-        qc: &'a QuorumCert,
+    async fn retrieve_block_for_id(
+        &mut self,
+        block_id: HashValue,
+        peers: &mut Vec<&AccountAddress>,
         num_blocks: u64,
     ) -> anyhow::Result<Vec<Block>> {
-        let block_id = qc.certified_block().id();
-        let mut peers: Vec<&AccountAddress> = qc.ledger_info().signatures().keys().collect();
         let mut attempt = 0_u32;
-        loop {
-            if peers.is_empty() {
-                bail!(
-                    "Failed to fetch block {} in {} attempts: no more peers available",
-                    block_id,
-                    attempt
-                );
-            }
-            let peer = self.pick_peer(attempt, &mut peers);
+        let mut progress = 0;
+        let mut last_block_id = block_id;
+        let mut result_blocks: Vec<Block> = vec![];
+        let mut retrieve_batch_size = min(num_blocks, MAX_BLOCKS_PER_REQUEST);
+        if peers.is_empty() {
+            bail!(
+                "Failed to fetch block {} in {} attempts: no more peers available",
+                block_id,
+                attempt
+            );
+        }
+        let mut peer = self.pick_peer(attempt, peers);
+        while progress < num_blocks {
+            // in case this is the last retrieval
+            retrieve_batch_size = min(retrieve_batch_size, num_blocks - progress);
+
             attempt += 1;
 
             debug!(
                 LogSchema::new(LogEvent::RetrieveBlock).remote_peer(peer),
                 block_id = block_id,
-                "Fetching block, attempt {}",
+                "Fetching {} blocks, attempt {}",
+                retrieve_batch_size,
                 attempt
             );
             let response = self
                 .network
                 .request_block(
-                    BlockRetrievalRequest::new(block_id, num_blocks),
+                    BlockRetrievalRequest::new(last_block_id, retrieve_batch_size),
                     peer,
                     retrieval_timeout(attempt),
                 )
                 .await;
             match response.and_then(|result| {
-                if result.status() == BlockRetrievalStatus::Succeeded {
+                if result.status() == BlockRetrievalStatus::Succeeded && !result.blocks().is_empty()
+                {
                     Ok(result.blocks().clone())
                 } else {
                     Err(format_err!("{:?}", result.status()))
                 }
             }) {
-                result @ Ok(_) => return result,
-                Err(e) => warn!(
-                    remote_peer = peer,
-                    block_id = block_id,
-                    error = ?e, "Failed to fetch block, trying another peer",
-                ),
+                Ok(batch) => {
+                    // extend the result blocks
+                    progress += batch.len() as u64;
```

Comment:
do we need progress? just use result.len()?

Reply:
I used `progress` mainly because it is u64. Shall we change num_blocks to usize?